src/packages/utils/src/things/logger.ts

```diff
@@ -0,0 +1,70 @@
+import { createWriteStream } from "fs";
+import { format } from "util";
+export type LogFunc = (message?: any, ...optionalParams: any[]) => void;
+
+export type Logger = {
+  log: LogFunc;
+};
+
+export type InternalLogger = Logger & {
+  close: () => Promise<void>;
+};
+
+type LoggerConfig = {
+  baseLogger: Logger;
+  file?: number;
+};
+
+export function createLogger(config: LoggerConfig): InternalLogger {
+  const baseLog = (...params: any[]) => config.baseLogger.log(...params);
+
+  if ("file" in config && config.file !== undefined) {
+    if (typeof config.file !== "number") {
+      throw new Error(
+        `'config.file' was not correctly normalized to a file descriptor. This should not happen. Value: ${
+          config.file
+        } of type ${typeof config.file}`
+      );
+    }
+    const fd = config.file;
+
+    const diskLogFormatter = (message: any) => {
+      // trailing space after date is delimiter between date and message
+      const linePrefix = `${new Date().toISOString()} `;
+      return message.toString().replace(/^/gm, linePrefix);
+    };
+
+    const writeStream = createWriteStream(null, { fd });
+
+    const onError = err =>
+      console.error(`Error writing to log file: ${err.message}`);
+    writeStream.on("error", onError);
+
+    const log = (message: any, ...optionalParams: any[]) => {
+      // we are logging to a file, but we still need to writing to console
+      baseLog(message, ...optionalParams);
+
+      const formattedMessage: string = format(message, ...optionalParams);
```

Comment:
A thought crossed my mind, i'm not sure if it is a good thought. But here is an alternative to ponder:

```suggestion
    const log = (...params: any[]) => {
      // we are logging to a file, but we still need to write to console
      baseLog(...params);

      const formattedMessage: string = format(...params);
```

Reply:
I actually prefer having the full signature, because this is more explicit, and easier to understand how the different functions with the same signature are used together. I'm happy to hear alternative viewpoints though if you or anyone else disagrees.

I updated another signature within this file that had the `(...params)` signature to include the `optionalParams` parameter as well.